field/src/coset.rs

```diff
@@ -0,0 +1,222 @@
+use core::iter::Take;
+
+use crate::{Powers, TwoAdicField};
+
+/// Coset of a subgroup of the group of units of a finite field of order equal
+/// to a power of two.
+///
+/// # Examples
+///
+/// ```
+/// # use p3_field::{
+///     TwoAdicField,
+///     PrimeCharacteristicRing,
+///     coset::TwoAdicMultiplicativeCoset
+/// };
+/// # use itertools::Itertools;
+/// # use p3_baby_bear::BabyBear;
+/// #
+/// type F = BabyBear;
+/// let log_size = 3;
+/// let shift = F::from_u64(7);
+/// let mut coset = TwoAdicMultiplicativeCoset::new(shift, log_size).unwrap();
+/// let generator = coset.subgroup_generator();
+///
+/// // Coset elements can be queried by index
+/// assert_eq!(coset.element(4), shift * generator.exp_u64(4));
+///
+/// // Coset elements can be iterated over in the canonical order
+/// assert_eq!(
+///     coset.iter().collect_vec(),
+///     (0..1 << log_size).map(|i| shift * generator.exp_u64(i)).collect_vec()
+/// );
+///
+/// // Cosets can be (element-wise) raised to a power of 2, either maintaining
+/// // the shift and raising only the subgroup, or raising both.
+/// let coset_shrunk_subgroup = coset.shrink_coset(2).unwrap();
+/// assert_eq!(
+///     coset_shrunk_subgroup.subgroup_generator(),
+///     coset.subgroup_generator().exp_power_of_2(2),
+/// );
+/// assert_eq!(
+///     coset_shrunk_subgroup.shift(),
+///     coset.shift()
+/// );
+///
+/// let coset_power = coset.exp_power_of_2(2).unwrap();
+/// assert_eq!(
+///     coset_power.subgroup_generator(),
+///     coset.subgroup_generator().exp_power_of_2(2),
+/// );
+/// assert_eq!(
+///     coset_power.shift(),
+///     coset.shift().exp_power_of_2(2),
+/// );
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct TwoAdicMultiplicativeCoset<F: TwoAdicField> {
+    // Letting s = shift, and g = generator (of order 2^log_size), the coset in
+    // question is
+    //     s * <g> = {s, s * g, shift * g^2, ..., s * g^(2^log_size - 1)]
+    shift: F,
+    log_size: usize,
+}
+
+impl<F: TwoAdicField> TwoAdicMultiplicativeCoset<F> {
+    /// Returns the coset `shift * <generator>`, where `generator` is a
+    /// canonical (i. e. fixed in the implementation of `F: TwoAdicField`)
+    /// generator of the unique subgroup of the units of `F` of order `2 ^
+    /// log_size`. Returns `None` if `log_size > F::TWO_ADICITY`.
+    ///
+    /// # Arguments
+    ///
+    ///  - `shift`: the value by which the subgroup is (multiplicatively)
+    ///    shifted
+    ///  - `log_size`: the size of the subgroup (and hence of the coset) is `2 ^
+    ///    log_size`. This determines the subgroup uniquely.
+    pub fn new(shift: F, log_size: usize) -> Option<Self> {
+        if log_size <= F::TWO_ADICITY {
+            Some(Self { shift, log_size })
+        } else {
+            None
+        }
+    }
+
+    /// Returns the generator of the subgroup whose shift is `self`.
```

Comment:
I'm not quite sure what you mean by: the subgroup whose shift is `self`.
```suggestion
    /// Returns the generator of the subgroup of order `self.size()`.
```

Reply:
The thinking behind the original comment was: `self` is a coset of the form `s * <g>`, so it is a shift of the subgroup `<g>` (of the group of units of the field).

At any rate, @Cesar199999 also got confused when he read that original comment so it's clear my wording wasn't the best. Happy to switch to your proposed wording (already committed) 👍